ably/proto/message.go

```diff
@@ -22,12 +27,273 @@ type Message struct {
 	ClientID     string                 `json:"clientId,omitempty" codec:"clientId,omitempty"`
 	ConnectionID string                 `json:"connectionId,omitempty" codec:"connectionID,omitempty"`
 	Name         string                 `json:"name,omitempty" codec:"name,omitempty"`
-	Data         string                 `json:"data,omitempty" codec:"data,omitempty"`
+	Data         *DataValue             `json:"data,omitempty" codec:"data,omitempty"`
```

Comment:
As it currently stands, the exposed API around `Message` is a bit clunky and it leaks implementation details to the outside. The fact that it needs a `*DataValue` means that you can't just instantiate them with e.g. `&Message{Data: []byte{1,2,3}}`, and since there's no "constructor" function, instantiating messages is harder than it should be. The `DataValue` type, the various "internal" but public methods etc. are implementation details that users shouldn't have to deal with.

So as before, my suggestion would be:

-  have an exported `Message` type that has `Data interface{}`, with a simple public API for creating new messages and working with them, and no exported "internal" methods. 

- have an `InternalMessage` (or whatever) type (this can even be as simple as `type InternalMessage proto.Message` + methods for `InternalMessage` that convert the data to a `DataValue` etc.) in an internal package under `ably/internal` that provides all the needed shared functionality that we don't want to expose an API for (so marshaling/unmarshaling, encoding/decoding etc.), but need to have shared in different parts of the library itself. It would also need easy conversions from/to `InternalMessage`. All public-facing APIs should only return the exported "dumb" `Message` type

Reply:
Note that ^ isn't 100% necessary to do in this PR, but I believe it really needs to be done eventually